# The Mechanical Lifecycle of a Token

The Mechanical Lifecycle of a Token: A Conceptual Primer on AI Language
1. The Foundation: Resetting Your Mental Model
To architect effective AI systems, you must first perform a radical mental model reset. Most users fail because they interact with Large Language Models (LLMs) as if they were high-speed search engines or reasoning entities. In reality, you are transitioning from Deterministic Systems—traditional software where logic is hard-coded—to Probabilistic Systems.

The following table illustrates the structural divide that every AI Architect must internalize:

- Dimension	| Deterministic Systems | (SQL, Calculators)	| Probabilistic Systems (LLMs)
- Input-Output Relationship | Same input → Same output (Repeatable)  | 	Outputs are distributions (Stochastic)
- Core Logic	Hard-coded rules and logic gates	Statistical likelihood and pattern recognition
- Primary Goal	Verifiable correctness and exact results	Fluency and likely continuations

**The "So What?" :**

- Treating an LLM like a search engine is a primary cause of system failure. Because these models optimize for fluency (sounding right) rather than truth (being right), they lack a native concept of "fact."
- To build reliable tools, you must stop "talking to a mind" and start "navigating a probability space."
- To understand these probability engines, we must first look at the "physical" units they process.
--------------------------------------------------------------------------------
2. Tokenization: How Machines "Read" Without Words
Machines do not see letters or words; they process numerical identifiers (IDs). Tokenization is the mechanical bridge that translates human language into these numbers. Most modern models use Byte Pair Encoding (BPE), which breaks text into subword units. This allows the model to handle a vast vocabulary while breaking rare or complex words into common fragments.
Consider the transformation of the phrase " Architecting":
Input Text: " Architecting"
1. Subword Tokens: [" Arch", "it", "ecting"]
2. Numerical IDs:  [4321, 85, 962]

Concept Spotlight:
- The Butterfly Effect of Tokens Because the machine's entire statistical path is based on these IDs, a single space or a capital letter (e.g., " Architect" vs. "architect") completely changes the numerical input. In this environment, a space is not "empty"—it is a distinct data point that shifts the model’s entire probability trajectory.
Once language is reduced to these disparate numbers, the engine must organize them into a map of "meaning."
--------------------------------------------------------------------------------
3. Embeddings: Visualizing Meaning as High-Dimensional Geometry
After tokenization, the model maps these IDs into Embeddings—coordinates in a high-dimensional mathematical space. In this world, Meaning = Geometry. The model uses Semantic Proximity to "understand" the world; if "Doctor" and "Hospital" frequently co-occur in training, their vectors are placed closer together.
You must internalize these Three Critical Misconceptions:
	• Embeddings are NOT Knowledge: They represent statistical proximity, not an understanding of the world.
	• Embeddings are NOT Facts: They indicate pattern similarity, not verified truth.
	• Embeddings are NOT Memory: They are static mathematical representations of language patterns.

Why This Matters for Architects

- Because embeddings are static representations of training data, they cannot represent real-time updates or "know" new information. This is why we utilize Vector Databases to store external knowledge and GraphRAG to preserve complex relationships. Distance in space is the machine's substitute for a world model, and as an architect, you must provide the "map" (grounding data) that the model lacks.
Once the map is drawn, the engine must decide how to move through it.
--------------------------------------------------------------------------------
4. The Engine: The Autoregressive Loop of Next-Token Prediction
The core of an LLM is a prediction engine governed by the formula P(next token | context). Much like an advanced autocomplete keyboard, the model looks at the current string of text and guesses the most likely next token.
The Three Pillars of the Prediction Engine:
	1. Probability: The model selects tokens based on a distribution of likelihoods.
	2. Context Accumulation: Every token the model predicts is appended back into the context to inform the next prediction.
	3. Pattern Recognition: The engine prioritizes completing a pattern (like a list or a grammatical structure) over the accuracy of the data itself.

The Danger of Error Compounding

- The model functions through Autoregressive Generation, meaning it predicts one token at a time. If the model selects one statistically "likely" but incorrect token, that error becomes part of the permanent context for all future steps. This shifts the probability distribution, often leading the model into a "logical drift" where it moves further away from facts as the generation continues.
We can, however, exert control over the "creativity" or "stability" of these predictions.
--------------------------------------------------------------------------------
5. Navigation Controls: The Science of the "Guess" (Sampling)
We use sampling parameters to navigate the model's probability distribution, balancing the trade-off between "Greedy" (deterministic-like) and "Creative" (random) outputs.

    • Temperature: This controls the "entropy" of the output.
    - Low Temp (0.1–0.4): Makes the model "greedy" and stable. It consistently picks the highest-probability tokens.
    - High Temp (0.7–1.0+): Flattens the distribution, allowing the model to pick "long-tail" tokens, which increases creativity but raises the risk of nonsense.

    • Top-P (Nucleus Sampling): This prunes the search space by only considering the top tokens whose cumulative probability mass adds up to value P. This effectively cuts off the "tail noise" of highly unlikely words.

Sampling Effects in Practice

	• Writing Code (Low Temp/Low P): Requires high stability and adherence to rigid logic.

	• Brainstorming (High Temp/High P): Requires a wider search of the probability space to generate novel connections.
These probabilistic guesses are the direct cause of the phenomenon known as "hallucination."
--------------------------------------------------------------------------------
6. Structural Reality: Why Hallucinations are Not "Bugs"
In traditional software, a bug is a failure of logic. In Generative AI, a hallucination—a statistically plausible but ungrounded output—is a natural byproduct of the architecture.
Anchor Truth: Hallucinations are structural, not mistakes to be "fixed." They are the fundamental result of a machine following probability rather than truth.
A major driver of this is Frequency Bias: the model will prioritize a "popular" wrong answer from its training data over a "rare" truth. Hallucinations predictably increase in these three scenarios:
	1. Ambiguity: Unclear prompts force the model to "fill the gap" with high-probability generic patterns.
	2. Missing Context: Without specific data to ground it, the model defaults to the most frequent internet patterns (Frequency Bias).
	3. Long Generations: As error compounding sets in, the model drifts further from the original intent.
Understanding these limitations shifts the responsibility from the model to the human designer.
--------------------------------------------------------------------------------
7. Conclusion: From Prompt User to AI Architect
The journey from tokenization to generation reveals a vital truth: Generative AI is not thinking; it is navigating a probability space. To succeed as an AI Architect, you must accept that the model has no native concept of truth. Intelligence must be engineered externally through grounding, constraints, and architecture.
Takeaway Checklist
	• [ ] Reset the Model: Stop treating LLMs like search engines; they are probability engines that optimize for fluency, not truth.
	• [ ] Respect the Token: Internalize that formatting, casing, and spaces change the machine's statistical trajectory.
	• [ ] Mind the Geometry: Use embeddings for semantic similarity, but remember they are static and require Vector Databases for real-time relevance.
	• [ ] Control the Sampling: Use Temperature and Top-P to prune the cumulative probability mass based on the specific task.
	• [ ] Architect for Truth: Since hallucinations are structural, you must engineer correctness through RAG, constraints, and evaluation.


The Ultimate Closing Mental Model: Intelligence is not found within the model alone it is engineered into the system by those who understand its mechanical lifecycle.


    1. Why tokenization exists (theoretical foundation)
    2. Algorithm families (BPE, SentencePiece, WordPiece, etc.)
    3. Framework ecosystem (libraries + model-specific tokenizers)
    4. Production considerations (speed, memory, multilingual, cost)
    5. Failure modes & edge cases
    6. How to teach this as a system-level concept

# Implementing BPE Algorithm


Implementing BPE
Now let’s take a look at an implementation of the BPE algorithm. This won’t be an optimized version you can actually use on a big corpus; we just want to show you the code so you can understand the algorithm a little bit better.

First we need a corpus, so let’s create a simple one with a few sentences:

**Learning Outcomes**

Students will be able to:

Extract clean text from PDFs

    Build a line-based training corpus
    Train a tokenizer (BPE + SentencePiece)
    Measure token inflation, compression ratio, and OOV behavior
    Understand why tokenizer choices affect context windows + cost

In [ ]:
corpus = [
    "This is the Implementing BPE.",
    "This section is about tokenization.",
    "This section shows several tokenizer algorithms.",
    "Hopefully, it will help you to understand how they are trained and generate tokens.",
]

- This code defines a Python list named corpus. This list contains four strings, each representing a sentence. This corpus will be used as a sample text dataset for the subsequent steps in the notebook, particularly for training and demonstrating the tokenization algorithms like BPE.

In [ ]:
# pip install ipywidgets

Next, we need to pre-tokenize that corpus into words. Since we are replicating a BPE tokenizer (like GPT-2), we will use the gpt2 tokenizer for the pre-tokenization:

In [ ]:
from transformers import AutoTokenizer

tokenizer = AutoTokenizer.from_pretrained("gpt2")

This code imports the AutoTokenizer class from the transformers library, which is a powerful tool for working with pre-trained tokenizers from Hugging Face. It then initializes a tokenizer by loading the pre-trained 'gpt2' tokenizer. This tokenizer object can then be used to convert text into numerical tokens that a GPT-2 model can understand.

In [ ]:
print(type(tokenizer))

<class 'transformers.models.gpt2.tokenization_gpt2.GPT2Tokenizer'>


In [ ]:
tokens = tokenizer("Hello world")
print(tokens)

{'input_ids': [15496, 995], 'attention_mask': [1, 1]}


This code snippet uses the tokenizer (which was loaded as a GPT-2 tokenizer) to process the string "Hello world".


---

- tokens = tokenizer("Hello world"): This line takes the input string "Hello world" and converts it into a format that a language model can understand. The tokenizer breaks the text into tokens (subword units) and then maps these tokens to numerical IDs.

- print(tokens): This line prints the output of the tokenization, which is typically a dictionary containing:
  - 'input_ids': A list of integers, where each integer is the unique ID for a corresponding token. For "Hello world", it shows [15496, 995].
  - 'attention_mask': A list of 1s and 0s indicating which tokens should be attended to by the model. A 1 means the token is part of the actual sequence, and 0 might be used for padding if multiple sequences were processed together. In this case, [1, 1] means both tokens are active.

In [ ]:
from collections import defaultdict

word_freqs = defaultdict(int)

for text in corpus:
    words_with_offsets = tokenizer.backend_tokenizer.pre_tokenizer.pre_tokenize_str(text)
    new_words = [word for word, offset in words_with_offsets]
    for word in new_words:
        word_freqs[word] += 1

print(word_freqs)

defaultdict(<class 'int'>, {'This': 3, 'Ġis': 2, 'Ġthe': 1, 'ĠImplementing': 1, 'ĠBPE': 1, '.': 4, 'Ġsection': 2, 'Ġabout': 1, 'Ġtokenization': 1, 'Ġshows': 1, 'Ġseveral': 1, 'Ġtokenizer': 1, 'Ġalgorithms': 1, 'Hopefully': 1, ',': 1, 'Ġit': 1, 'Ġwill': 1, 'Ġhelp': 1, 'Ġyou': 1, 'Ġto': 1, 'Ġunderstand': 1, 'Ġhow': 1, 'Ġthey': 1, 'Ġare': 1, 'Ġtrained': 1, 'Ġand': 1, 'Ġgenerate': 1, 'Ġtokens': 1})


The 'Ġ' character (often referred to as 'G-dot' or 'G-space') is a common convention used in Byte Pair Encoding (BPE) tokenizers, especially those derived from or similar to GPT-2's tokenizer. It signifies a space character that precedes the token.

For example, if you tokenize the phrase " Hello world", the tokenizer might produce ['ĠHello', 'Ġworld']. Here:

ĠHello means the token is Hello (a space followed by "Hello").
Ġworld means the token is world (a space followed by "world").
This convention is crucial for:

Reconstruction: It allows the tokenizer to accurately reconstruct the original text, including spaces.
Handling Leading Spaces: It helps distinguish between a word at the beginning of a sentence (which might not have a leading space) and a word in the middle of a sentence (which usually does).
Without it, "Hello" and " Hello" would be tokenized identically, leading to ambiguity when decoding back to text.

- This code snippet calculates the frequency of each pre-tokenized word in your corpus. Let's break it down:

- from collections import defaultdict: Imports the defaultdict class, which is a subclass of dict that provides a default value for a key that has not been set.

- word_freqs = defaultdict(int): Initializes an empty defaultdict where the default value for any new key will be 0 (an integer).

- The for text in corpus: loop iterates through each sentence in your corpus.

- words_with_offsets = tokenizer.backend_tokenizer.pre_tokenizer.

- pre_tokenize_str(text): This is the crucial part. It uses the pre_tokenizer component of the loaded GPT-2 tokenizer to break down each sentence into words (or subword units) along with their starting offsets. The pre_tokenize_str method performs initial tokenization before BPE merging.

- new_words = [word for word, offset in words_with_offsets]:

- Extracts only the word/token strings from the words_with_offsets list, ignoring the offset information.

- The inner for word in new_words: loop iterates through these extracted words.

- word_freqs[word] += 1: For each word, it increments its count in the word_freqs dictionary. If the word is encountered for the first time, it's added to the dictionary with a count of 1 (thanks to defaultdict(int)).

- print(word_freqs): Finally, it prints the word_freqs dictionary, showing each unique pre-tokenized word and its total count across the entire corpus.

The next step is to compute the base vocabulary, formed by all the characters used in the corpus:

In [ ]:
alphabet = []

for word in word_freqs.keys():
    for letter in word:
        if letter not in alphabet:
            alphabet.append(letter)
alphabet.sort()

print(alphabet)

[',', '.', 'B', 'E', 'H', 'I', 'P', 'T', 'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'k', 'l', 'm', 'n', 'o', 'p', 'r', 's', 't', 'u', 'v', 'w', 'y', 'z', 'Ġ']


We also add the special tokens used by the model at the beginning of that vocabulary. In the case of GPT-2, the only special token is "<|endoftext|>":

This code snippet constructs an alphabet list containing all unique characters present across all the pre-tokenized words in your word_freqs dictionary. Here's how it works:

	• alphabet = []: Initializes an empty list that will store all the unique characters.

	• for word in word_freqs.keys():: It iterates through each unique pre-tokenized word (like "This", "Ġis", etc.) found in your corpus.

	• for letter in word:: For each word, it then iterates through every individual character (letter) within that word.

	• if letter not in alphabet:: This condition checks if the current letter is already present in the alphabet list. This ensures that only unique characters are added.

	• alphabet.append(letter): If the letter is not already in the alphabet list, it's added.

	• alphabet.sort(): After processing all words and characters, this line sorts the alphabet list alphabetically.

    • print(alphabet): Finally, it prints the complete, sorted list of unique characters.

In [ ]:
vocab = ["<|endoftext|>"] + alphabet.copy()

This line of code initializes the vocab list, which will serve as the starting vocabulary for the Byte Pair Encoding (BPE) algorithm. It does this by:

["<|endoftext|>"]: Starting with a list containing a special token, "<|endoftext|>". This token is commonly used in models like GPT-2 to signal the end of a text or sequence.
+ alphabet.copy(): Appending a copy of the alphabet list to it. The alphabet contains all the unique individual characters extracted from your corpus. Using .copy() ensures that alphabet itself is not modified if vocab changes later.
So, the vocab initially consists of this special token and all the individual characters, forming the most basic units that the tokenizer can recognize before any merges are learned.

We now need to split each word into individual characters, to be able to start training:

In [ ]:
splits = {word: [c for c in word] for word in word_freqs.keys()}

Now that we are ready for training, let’s write a function that computes the frequency of each pair. We’ll need to use this at each step of the training:

In [ ]:
def compute_pair_freqs(splits):
    pair_freqs = defaultdict(int)
    for word, freq in word_freqs.items():
        split = splits[word]
        if len(split) == 1:
            continue
        for i in range(len(split) - 1):
            pair = (split[i], split[i + 1])
            pair_freqs[pair] += freq
    return pair_freqs

This compute_pair_freqs function is crucial for the Byte Pair Encoding (BPE) algorithm. Its purpose is to calculate how often each adjacent pair of tokens (characters or already merged subwords) appears across all the words in your corpus, weighted by the frequency of each word. Let's break down how it works:

def compute_pair_freqs(splits):: Defines a function that takes one argument: splits, which is a dictionary mapping words to their current list of split tokens (e.g., {'Architecting': ['A', 'r', 'c', 'h', 'i', 't', 'e', 'c', 't', 'i', 'n', 'g']}).
pair_freqs = defaultdict(int): Initializes an empty defaultdict where keys will be token pairs (tuples like ('Ġ', 't')) and values will be their frequencies.
for word, freq in word_freqs.items():: It iterates through each unique word in the corpus and its overall frequency (word_freqs is a global variable holding the frequency of each word).
split = splits[word]: Retrieves the current tokenized representation (the 'split') of the word from the splits dictionary.
if len(split) == 1: continue: If a word has only one token (meaning no pairs can be formed), it skips to the next word.
for i in range(len(split) - 1):: This loop iterates through the split list, stopping one element before the end to ensure split[i + 1] is always valid.
pair = (split[i], split[i + 1]): Creates a tuple representing an adjacent pair of tokens.
pair_freqs[pair] += freq: Increments the frequency of this pair in the pair_freqs dictionary. The increment is by freq (the frequency of the original word) rather than just 1, ensuring that common words contribute more to the pair frequencies.
return pair_freqs: The function returns the pair_freqs dictionary, containing all unique adjacent token pairs and their aggregated frequencies.


Let’s have a look at a part of this dictionary after the initial splits:

In [ ]:
pair_freqs = compute_pair_freqs(splits)

for i, key in enumerate(pair_freqs.keys()):
    print(f"{key}: {pair_freqs[key]}")
    if i >= 5:
        break

('T', 'h'): 3
('h', 'i'): 3
('i', 's'): 5
('Ġ', 'i'): 3
('Ġ', 't'): 7
('t', 'h'): 3


Now, finding the most frequent pair only takes a quick loop:

This code first calculates the frequency of all possible adjacent token pairs using the compute_pair_freqs function, which we discussed earlier. After computing all the pair_freqs, the for loop then iterates through these pairs and their frequencies. The enumerate(pair_freqs.keys()) part assigns an index i to each key (pair).

It prints each pair along with its frequency in the format (token1, token2): frequency. The if i >= 5: break condition ensures that only the first 6 unique pairs and their frequencies are printed, providing a quick glance at some of the most common token combinations found.

In [ ]:
best_pair = ""
max_freq = None

for pair, freq in pair_freqs.items():
    if max_freq is None or max_freq < freq:
        best_pair = pair
        max_freq = freq

print(best_pair, max_freq)

('Ġ', 't') 7


This code block finds the most frequent pair from the pair_freqs dictionary. It initializes best_pair as an empty string and max_freq to None. Then, it iterates through each pair and its frequency in pair_freqs.

If max_freq is None (meaning it's the first pair being checked) or if the current freq is greater than the max_freq found so far, it updates best_pair to the current pair and max_freq to the current freq. Finally, it prints the best_pair and its max_freq.

So the first merge to learn is ('Ġ', 't') -> 'Ġt', and we add 'Ġt' to the vocabulary:

In [ ]:
merges = {("Ġ", "t"): "Ġt"}
vocab.append("Ġt")

This output, ('Ġ', 't') 7, means that the most frequent adjacent pair of tokens found in your corpus is ('Ġ', 't'), and it occurred with a total frequency of 7. According to the Byte Pair Encoding (BPE) algorithm, this is the pair that will be merged next to form a new, single token, Ġt, which will then be added to the vocabulary.

To continue, we need to apply that merge in our splits dictionary. Let’s write another function for this:

In [ ]:
def merge_pair(a, b, splits):
    for word in word_freqs:
        split = splits[word]
        if len(split) == 1:
            continue

        i = 0
        while i < len(split) - 1:
            if split[i] == a and split[i + 1] == b:
                split = split[:i] + [a + b] + split[i + 2 :]
            else:
                i += 1
        splits[word] = split
    return splits

This merge_pair function is designed to apply a learned BPE merge rule across all words in your corpus. It takes three arguments:

a and b: The two tokens that form the pair to be merged (e.g., "Ġ" and "t").
splits: The dictionary where keys are original words and values are their current tokenized representations (lists of tokens).
Here's how it works:

Iterate through words: It loops through each word in your corpus (using word_freqs to get the list of words).
Get current split: For each word, it retrieves its current split representation from the splits dictionary.
Skip single tokens: If a word is already represented by a single token, there are no pairs to merge, so it skips to the next word.
Find and Replace Loop: It then iterates through the split list for the current word.
If it finds the sequence a followed by b (split[i] == a and split[i + 1] == b), it merges them. It reconstructs the split list by taking everything before a, adding the new merged token a + b, and then adding everything after b.
If the pair is not found, it moves to the next position in the split list.
Update splits: After processing all possible merges for a word, the modified split list is stored back into the splits dictionary for that word.
The function ultimately returns the updated splits dictionary, reflecting the application of the specified merge rule across the entire corpus.

And we can have a look at the result of the first merge:

In [ ]:
splits = merge_pair("Ġ", "t", splits)
print(splits["Ġtrained"])

['Ġt', 'r', 'a', 'i', 'n', 'e', 'd']


This code first applies the merge_pair function. It merges the pair ('Ġ', 't') into a new token ('Ġt') across all applicable words in the splits dictionary. Then, it prints the tokenized representation of the word "Ġtrained" after this merge has been applied. You'll see how "Ġ" followed by "t" in "Ġtrained" has been replaced by the single token "Ġt".

Now we have everything we need to loop until we have learned all the merges we want. Let’s aim for a vocab size of 50:

In [ ]:
vocab_size = 50

while len(vocab) < vocab_size:
    pair_freqs = compute_pair_freqs(splits)
    best_pair = ""
    max_freq = None
    for pair, freq in pair_freqs.items():
        if max_freq is None or max_freq < freq:
            best_pair = pair
            max_freq = freq
    splits = merge_pair(*best_pair, splits)
    merges[best_pair] = best_pair[0] + best_pair[1]
    vocab.append(best_pair[0] + best_pair[1])

As a result, we’ve learned 19 merge rules (the initial vocabulary had a size of 31 — 30 characters in the alphabet, plus the special token):

This code block implements the core training loop for the Byte Pair Encoding (BPE) algorithm. It continues to run as long as the current vocab size is less than the desired vocab_size (which is set to 50):

pair_freqs = compute_pair_freqs(splits): In each iteration, it first re-calculates the frequencies of all possible adjacent token pairs based on the current state of the splits dictionary.
Find Best Pair: It then finds the best_pair (the one with the max_freq) among all the newly computed pair_freqs.
splits = merge_pair(*best_pair, splits): The merge_pair function is called to apply this newly found best_pair merge across all words in the splits dictionary, effectively replacing occurrences of the best_pair with its merged single token.
Update Merges and Vocab: The best_pair and its merged string representation are added to the merges dictionary, and the new merged token is appended to the vocab list. This expands the vocabulary with a new, longer subword unit.
This loop continues, gradually building up the vocab with increasingly complex subword tokens by merging the most frequent adjacent pairs, until the vocab_size target is met.

In [ ]:
print(merges)

{('Ġ', 't'): 'Ġt', ('i', 's'): 'is', ('e', 'n'): 'en', ('t', 'i'): 'ti', ('Ġ', 's'): 'Ġs', ('Ġ', 'a'): 'Ġa', ('Ġt', 'o'): 'Ġto', ('e', 'r'): 'er', ('T', 'h'): 'Th', ('Th', 'is'): 'This', ('h', 'e'): 'he', ('Ġs', 'e'): 'Ġse', ('ti', 'o'): 'tio', ('tio', 'n'): 'tion', ('Ġto', 'k'): 'Ġtok', ('Ġtok', 'en'): 'Ġtoken', ('n', 'd'): 'nd'}


And the vocabulary is composed of the special token, the initial alphabet, and all the results of the merges:

In [ ]:
print(vocab)

['<|endoftext|>', ',', '.', 'B', 'E', 'H', 'I', 'P', 'T', 'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'k', 'l', 'm', 'n', 'o', 'p', 'r', 's', 't', 'u', 'v', 'w', 'y', 'z', 'Ġ', 'Ġt', 'is', 'en', 'ti', 'Ġs', 'Ġa', 'Ġto', 'er', 'Th', 'This', 'he', 'Ġse', 'tio', 'tion', 'Ġtok', 'Ġtoken', 'nd']


To tokenize a new text, we pre-tokenize it, split it, then apply all the merge rules learned:

In [ ]:
def tokenize(text):
    pre_tokenize_result = tokenizer._tokenizer.pre_tokenizer.pre_tokenize_str(text)
    pre_tokenized_text = [word for word, offset in pre_tokenize_result]
    splits = [[l for l in word] for word in pre_tokenized_text]
    for pair, merge in merges.items():
        for idx, split in enumerate(splits):
            i = 0
            while i < len(split) - 1:
                if split[i] == pair[0] and split[i + 1] == pair[1]:
                    split = split[:i] + [merge] + split[i + 2 :]
                else:
                    i += 1
            splits[idx] = split

    return sum(splits, [])

We can try this on any text composed of characters in the alphabet:

This tokenize function is designed to take a raw text string and convert it into a list of tokens using the BPE merges learned previously. Here's a breakdown of its steps:

Initial Pre-tokenization: pre_tokenize_result = tokenizer._tokenizer.pre_tokenizer.pre_tokenize_str(text): First, it applies the initial pre-tokenization step using the GPT-2 tokenizer's internal pre-tokenizer. This breaks the input text into words or basic subword units and their offsets.
Extract Words: pre_tokenized_text = [word for word, offset in pre_tokenize_result]: It then extracts just the word strings from the pre-tokenization result.
Initial Splits to Characters: splits = [[l for l in word] for word in pre_tokenized_text]: Each of these pre-tokenized words is further split into individual characters. At this point, splits is a list of lists, where each inner list contains the characters of a word (e.g., [['T', 'h', 'i', 's'], ['i', 's'], ...]).
Apply BPE Merges: The function then iterates through all the pair and merge rules that were learned and stored in the merges dictionary.
For each learned merge rule, it iterates through every split (representing a word) in the splits list.
Inside this loop, it searches for occurrences of the pair (split[i] == pair[0] and split[i + 1] == pair[1]).
When a match is found, the two tokens pair[0] and pair[1] are replaced by their merge (e.g., ('Ġ', 't') becomes 'Ġt'). This process continues for all possible merges within each word until no more merges for that pair can be applied.
Flatten Results: return sum(splits, []): Finally, sum(splits, []) is used to flatten the splits (which is a list of lists of tokens) into a single, cohesive list of tokens, representing the tokenized version of the original input text.

In [ ]:
tokenize("This is not a token.")

['This', 'Ġ', 'is', 'Ġ', 'n', 'o', 't', 'Ġa', 'Ġtoken', '.']

The tokenize function uses the merges dictionary to progressively combine individual characters or smaller subword units into larger, more meaningful subword tokens. After an initial pre-tokenization step that breaks the input text into words and then into individual characters, the function iterates through each pair and its corresponding merge value stored in the merges dictionary. For each such learned merge rule, it scans through the character-level representation of every word. Whenever it finds an occurrence of the pair (e.g., ('Ġ', 't')), it replaces those two tokens with their single merged representation (e.g., 'Ġt'). This process is repeated for all learned merges, gradually transforming the character-level splits into the final BPE-tokenized output.

The output of tokenize("This is not a token.") is ['This', 'Ġ', 'is', 'Ġ', 'n', 'o', 't', 'Ġa', 'Ġtoken', '.']. This result shows how the input text is broken down into tokens using the BPE merges that were learned previously. For example, 'This' was merged, 'Ġis' was not (as 'is' might be a common word), 'Ġa' and 'Ġtoken' were also formed by merging characters into subword units, and 'n', 'o', 't', and '.' remain as individual tokens, indicating that no frequent merges were found for them in this context or that they are considered base vocabulary characters.

In [ ]:
long_text = "This is a longer piece of text to test the tokenization. It includes several sentences and different words to observe how the BPE algorithm handles it."
tokenized_long_text = tokenize(long_text)
print(tokenized_long_text)
print(f"Number of tokens: {len(tokenized_long_text)}")

['This', 'Ġ', 'is', 'Ġa', 'Ġ', 'l', 'o', 'n', 'g', 'er', 'Ġ', 'p', 'i', 'e', 'c', 'e', 'Ġ', 'o', 'f', 'Ġt', 'e', 'x', 't', 'Ġto', 'Ġt', 'e', 's', 't', 'Ġt', 'he', 'Ġtoken', 'i', 'z', 'a', 'tion', '.', 'Ġ', 'I', 't', 'Ġ', 'i', 'n', 'c', 'l', 'u', 'd', 'e', 's', 'Ġse', 'v', 'er', 'a', 'l', 'Ġs', 'en', 't', 'en', 'c', 'e', 's', 'Ġa', 'nd', 'Ġ', 'd', 'i', 'f', 'f', 'er', 'en', 't', 'Ġ', 'w', 'o', 'r', 'd', 's', 'Ġto', 'Ġ', 'o', 'b', 's', 'er', 'v', 'e', 'Ġ', 'h', 'o', 'w', 'Ġt', 'he', 'Ġ', 'B', 'P', 'E', 'Ġa', 'l', 'g', 'o', 'r', 'i', 't', 'h', 'm', 'Ġ', 'h', 'a', 'nd', 'l', 'e', 's', 'Ġ', 'i', 't', '.']
Number of tokens: 114


We have text data (corpus).

We pre-tokenize into word-like chunks.

We count chunk frequency.

We extract all unique characters → alphabet.

We create base vocab = special tokens + alphabet.

We split each chunk into characters.

We repeatedly:

    count adjacent pair frequencies

    pick best pair

    merge it everywhere

    store the rule

    grow vocab

To tokenize new text, we:

    pre-tokenize

    split into characters

    replay merges in order

That is BPE.

#Part 0 — Setup

Create the below file paths:

  - /content/lab_tokenizer/data_pdfs/doc1.pdf

  - /content/lab_tokenizer/data_pdfs/doc2.pdf

In [ ]:
pip install tokenizers sentencepiece pypdf tqdm

# Part 1 — Extract Text from PDFs

In [ ]:
from pathlib import Path
from pypdf import PdfReader
from tqdm import tqdm
import re

PDF_DIR = Path("/content/lab_tokenizer/data_pdfs")
OUT_DIR = Path("/content/outputs")
OUT_DIR.mkdir(parents=True, exist_ok=True)

def clean_text(text: str) -> str:
    # Normalize whitespace and remove repeated spaces
    text = text.replace("\u00a0", " ")  # non-breaking space
    text = re.sub(r"[ \t]+", " ", text)
    text = re.sub(r"\n{3,}", "\n\n", text)
    return text.strip()

def pdf_to_text(pdf_path: Path) -> str:
    reader = PdfReader(str(pdf_path))
    parts = []
    for page in reader.pages:
        page_text = page.extract_text() or ""
        parts.append(page_text)
    return clean_text("\n".join(parts))

def split_to_lines(text: str, min_len: int = 20):
    # For tokenizer training, line-based corpus is best
    lines = [ln.strip() for ln in text.splitlines()]
    # keep meaningful lines only
    return [ln for ln in lines if len(ln) >= min_len]

def main():
    corpus_lines = []
    pdfs = sorted(PDF_DIR.glob("*.pdf"))

    if not pdfs:
        raise SystemExit(f"No PDFs found in {PDF_DIR.resolve()}")

    for pdf in tqdm(pdfs, desc="Extracting PDFs"):
        text = pdf_to_text(pdf)
        lines = split_to_lines(text)
        corpus_lines.extend(lines)

    out_file = OUT_DIR / "corpus.txt"
    out_file.write_text("\n".join(corpus_lines), encoding="utf-8")

    print(f"Corpus saved: {out_file} ({len(corpus_lines)} lines)")

if __name__ == "__main__":
    main()

Extracting PDFs: 100%|██████████| 2/2 [00:04<00:00,  2.03s/it]

Corpus saved: /content/outputs/corpus.txt (1452 lines)


# Part 2 — Train a BPE Tokenizer (HuggingFace tokenizers)

In [ ]:
from tokenizers import Tokenizer
from tokenizers.models import BPE
from tokenizers.trainers import BpeTrainer
from tokenizers.pre_tokenizers import ByteLevel
from tokenizers.decoders import ByteLevel as ByteLevelDecoder
from tokenizers.normalizers import NFKC
from pathlib import Path

CORPUS = Path("outputs/corpus.txt")
OUT_DIR = Path("outputs/bpe")
OUT_DIR.mkdir(parents=True, exist_ok=True)

def main():
    if not CORPUS.exists():
        raise SystemExit("Run 01_extract_pdf_text.py first.")

    tokenizer = Tokenizer(BPE(unk_token="[UNK]"))
    tokenizer.normalizer = NFKC()
    tokenizer.pre_tokenizer = ByteLevel(add_prefix_space=True)

    trainer = BpeTrainer(
        vocab_size=8000,
        min_frequency=2,
        special_tokens=["[UNK]", "[PAD]", "[BOS]", "[EOS]"]
    )

    tokenizer.train([str(CORPUS)], trainer)
    tokenizer.decoder = ByteLevelDecoder()

    tokenizer.save(str(OUT_DIR / "tokenizer.json"))
    print(f" Saved BPE tokenizer to {OUT_DIR / 'tokenizer.json'}")

if __name__ == "__main__":
    main()

 Saved BPE tokenizer to outputs/bpe/tokenizer.json


# Part 3 — Train a SentencePiece Tokenizer (LLaMA/T5 style)

In [ ]:
from pathlib import Path
import subprocess
import sentencepiece as spm # Import sentencepiece

CORPUS = Path("outputs/corpus.txt")
OUT_DIR = Path("outputs/sp")
OUT_DIR.mkdir(parents=True, exist_ok=True)

def main():
    if not CORPUS.exists():
        raise SystemExit("Run 01_extract_pdf_text.py first.")

    # Construct parameters as a single string for SentencePieceTrainer.train
    params = (
        f"--input={CORPUS} "
        f"--model_prefix={OUT_DIR / 'sp_unigram'} "
        "--vocab_size=4096 "
        "--model_type=unigram "
        "--character_coverage=1.0"
    )

    # Use the Python API for SentencePiece training
    spm.SentencePieceTrainer.train(params)
    print(f"Saved SentencePiece model: {OUT_DIR / 'sp_unigram.model'}")

if __name__ == "__main__":
    main()

Saved SentencePiece model: outputs/sp/sp_unigram.model


# Part 4 — Evaluate Tokenizer Quality (Token Inflation + Reconstruction)

In [ ]:
from pathlib import Path
import statistics
import sentencepiece as spm
from tokenizers import Tokenizer

CORPUS = Path("outputs/corpus.txt")

BPE_PATH = Path("outputs/bpe/tokenizer.json")
SP_PATH = Path("outputs/sp/sp_unigram.model")

TEST_SENTENCES = [
    "Architecting systems requires careful design.",
    " Architecting systems requires careful design.",  # leading space
    "Apple vs apple changes meaning.",
    "Why is the sky blue?",
    "JSON: {\"a\": 1, \"b\": 2}",
    "print('hello world')",
]

def load_lines(max_lines=2000):
    lines = CORPUS.read_text(encoding="utf-8").splitlines()
    # sample first max_lines for quick lab
    return [ln.strip() for ln in lines[:max_lines] if ln.strip()]

def bpe_stats(lines):
    tok = Tokenizer.from_file(str(BPE_PATH))
    token_counts = []
    char_counts = []

    for ln in lines:
        enc = tok.encode(ln)
        token_counts.append(len(enc.ids))
        char_counts.append(len(ln))

    inflation = [t/c if c else 0 for t, c in zip(token_counts, char_counts)]
    return {
        "avg_tokens_per_line": statistics.mean(token_counts),
        "avg_chars_per_line": statistics.mean(char_counts),
        "avg_tokens_per_char": statistics.mean(inflation),
        "avg_chars_per_token": statistics.mean([c/t if t else 0 for t, c in zip(token_counts, char_counts)]),
    }

def sp_stats(lines):
    sp = spm.SentencePieceProcessor()
    sp.load(str(SP_PATH))

    token_counts = []
    char_counts = []

    for ln in lines:
        ids = sp.encode(ln, out_type=int)
        token_counts.append(len(ids))
        char_counts.append(len(ln))

    inflation = [t/c if c else 0 for t, c in zip(token_counts, char_counts)]
    return {
        "avg_tokens_per_line": statistics.mean(token_counts),
        "avg_chars_per_line": statistics.mean(char_counts),
        "avg_tokens_per_char": statistics.mean(inflation),
        "avg_chars_per_token": statistics.mean([c/t if t else 0 for t, c in zip(token_counts, char_counts)]),
    }

def demo_examples():
    print("\n===== DEMO EXAMPLES =====")

    bpe = Tokenizer.from_file(str(BPE_PATH))
    sp = spm.SentencePieceProcessor()
    sp.load(str(SP_PATH))

    for s in TEST_SENTENCES:
        print("\nTEXT:", repr(s))

        bpe_enc = bpe.encode(s)
        print("BPE TOKENS:", bpe_enc.tokens)
        print("BPE DECODE:", bpe.decode(bpe_enc.ids))

        sp_pieces = sp.encode(s, out_type=str)
        sp_ids = sp.encode(s, out_type=int)
        print("SP PIECES:", sp_pieces)
        print("SP DECODE:", sp.decode(sp_ids))

def main():
    lines = load_lines()

    print("\n===== TOKENIZER METRICS (first ~2000 lines) =====")
    print("BPE:", bpe_stats(lines))
    print("SP :", sp_stats(lines))

    demo_examples()

if __name__ == "__main__":
    main()


===== TOKENIZER METRICS (first ~2000 lines) =====
BPE: {'avg_tokens_per_line': 21.495179063360883, 'avg_chars_per_line': 82.18939393939394, 'avg_tokens_per_char': 0.2702967084688243, 'avg_chars_per_token': 4.009598273550821}
SP : {'avg_tokens_per_line': 22.223140495867767, 'avg_chars_per_line': 82.18939393939394, 'avg_tokens_per_char': 0.27728611266625514, 'avg_chars_per_token': 3.853803249106052}

===== DEMO EXAMPLES =====

TEXT: 'Architecting systems requires careful design.'
BPE TOKENS: ['ĠAr', 'ch', 'it', 'ec', 'ting', 'Ġsystems', 'Ġrequires', 'Ġcareful', 'Ġdesign', '.']
BPE DECODE:  Architecting systems requires careful design.
SP PIECES: ['▁Ar', 'ch', 'ite', 'c', 'ting', '▁systems', '▁require', 's', '▁c', 'are', 'ful', '▁design', '.']
SP DECODE: Architecting systems requires careful design.

TEXT: ' Architecting systems requires careful design.'
BPE TOKENS: ['ĠAr', 'ch', 'it', 'ec', 'ting', 'Ġsystems', 'Ġrequires', 'Ġcareful', 'Ġdesign', '.']
BPE DECODE:  Architecting systems re

What students should interpret

avg_chars_per_token higher = better compression (usually)

avg_tokens_per_char lower = fewer tokens per context window (good)

Compare how leading spaces differ (BPE ByteLevel vs SentencePiece)

# Part 5 — Classroom Exercises (graded)

**Exercise 1 (Prompt cost)**

    Take a prompt and compute token count under both tokenizers.

    Which tokenizer is more token-efficient?

    What does this imply for context window + cost?

**Exercise 2 (Domain terms)**

    Create 15 domain words from the PDF (e.g., “Dataproc”, “Tokenization”, “KV-cache”, “microservices”)
    Evaluate how they split.

    Which tokenizer captures domain terms as single tokens faster?

    Why?

**Exercise 3 (Vocab size experiment)**

    Retrain with:

    vocab_size=4000

    vocab_size=16000

  Compare:

    token inflation ratio

    whether domain terms become single tokens

    Why does vocab size trade off embedding size vs sequence length?

**Exercise 4 (Noise sensitivity)**

  Add noisy lines:

    URLs

    emails

    random IDs

  Observe:

    Byte-level BPE robustness vs SentencePiece behavior

# Part 6 —  “Closing Insight”

Tokenizer determines:

    context window utilization
    compute per request
    model’s “perception resolution”
    multilingual efficiency
    scaling curves (because scaling measured in tokens)

So:

    Tokenization is not preprocessing.
    It is the language-to-tensor contract.